docs/ABI/GenericSignature.md

```diff
@@ -0,0 +1,164 @@
+
+* `T1` and `T2` are generic type parameters with depths `d1` and `d2`, and indices `i1` and `i2`, respectively, and either `d1 < d2` or `d1 == d2 && i1 < i2`;
+* `T1` is a generic type parameter and `T2` is a nested type `U2.A2`; or
+* `T1` is a nested type `U1.A1` and `T2` is a nested type `U2.A2`, where `A1` and `A2` name associated types of the protocols `P1` and `P2`, respectively, and either
+    * `U1` precedes `U2` in the canonical ordering, or
+    * `U1 == U2` and the name of `A1` lexicographically precedes the name of `A2`, or
+    * `U1 == U2` and `P1` precedes `P2` in the canonical ordering defined by the following section on [protocol ordering](#protocol-ordering).
+
+### Protocol ordering
+
+Given two protocols `P1` and `P2`, protocol `P1` precedes `P2` in the canonical ordering if:
+
+* `P1` is in a different module than `P2` and the module name of `P1` lexicographically precedes the module name of `P2`, or
+* `P1` and `P2` are in the same module and the name of `P1` lexicographically precedes the name of `P2`.
+
+### Canonical constraints
+
+A given constraint can be described in multiple ways. In our running example, the conformance constraint for the element type can be expressed as either `C1.Element: Equatable` or `C2.Element: Equatable`, because `C1.Element` and `C2.Element` name the same type. There might be an infinite number of ways to name the same type (e.g., `C1.SubSequence.SubSequence.Iterator.Element` is also equivalent to `C1.Element`). All of the spellings that refer to the same time comprise the *equivalence class* of that type.
+
+Each equivalence class has a corresponding *anchor*, which is a type parameter that is the least type according to the [type parameter ordering](#type-parameter-ordering). Anchors are used to describe requirements canonically. A concrete type (i.e., a type that is not a type parameter) is canonical when each type parameters within is the anchor of its equivalence class.
+
+A layout or conformance constraint is canonical when its left-hand side is the anchor of its equivalence class. A superclass constraint is canonical when its left-hand side is the anchor of its equivalence class and its right-hand side is a canonical concrete (class) type. Same-type constraint canonicalization is discussed in detail in the following section, but some basic rules apply: the left-hand side is always a type parameter, and the right-hand side is either a type parameter that follows the left-hand side (according to the [type parameter ordering](#type-parameter-ordering)) or is a canonical concrete type.
+
+### Same-type constaints
+
+The canonical form of superclass, layout, and conformance constraints are trivially canonicalized using the anchor of the appropriate equivalence class. Same-type constraints, on the other hand, are responsible for forming those equivalence classes. Let's expand our running example to include a third `Collection`:
```

Comment:
What about superclass constraints where the right hand side is a generic class?

Reply:
Yeah, I should call that out. The RHS needs to be a canonical concrete type.